R/pairwise-comparisons.R

```diff
@@ -0,0 +1,244 @@
+                               baseline = NULL) { # comparison against a baseline model yet to be implemented
+
+  # get list of models
+  identifiers <- unique(scores[[var]])
+
+  # identify unit of single observation
+  # this basically recovers what was used in eval_forecasts as by argument
+  # could think about adding a specific by argument to this function, but
+  # really this should not be necessary I think
+  all_metrics <- list_of_avail_metrics()
+  by <- setdiff(names(scores), c(all_metrics, var))
+
+  # for efficiency, remove everything that is not in c(by, var)
+  scores <- data.table::as.data.table(scores)
+  cols_to_remove <- setdiff(names(scores), c(by, var, metric))
+  scores[, eval(cols_to_remove) := NULL]
+
+
+  single_comparison <- function(model1, model2, permutation_test = FALSE) {
+    # select only columns in c(by, var)
+    a <- scores[model == model1, ]
+    b <- scores[model == model2, ]
+
+    overlap <- merge(a, b, by = by)
+
+    #-------------------------------------------------
+    # need to do error handling if there is no overlap.
+    #-------------------------------------------------
+
+    values_x <- overlap[[paste0(metric, ".x")]]
+    values_y <- overlap[[paste0(metric, ".y")]]
+
+    # ratio should be equivalent to theta_ij in Johannes document
+    ratio <- sum(values_x) / sum(values_y)
+
+    if (permutation_test) {
+      # taken from the surveillance package
+      permutation_test <- function(score1, score2, nPermutation = 999, ratio = FALSE) {
+        nTime = length(score1)
+        meanScore1 <- mean(score1)
+        meanScore2 <- mean(score2)
+
+        if (ratio) {
+          testStat_observed <- meanScore1 / meanScore2
+        } else {
+          testStat_observed <- meanScore1 - meanScore2
+        }
+
+        testStat_permuted <- replicate(nPermutation, {
+          sel <- rbinom(nTime, size = 1, prob = 0.5)
+          g1 <- (sum(score1[sel == 0]) + sum(score2[sel == 1]))/nTime
+          g2 <- (sum(score1[sel == 1]) + sum(score2[sel == 0]))/nTime
+
+          if (ratio) {
+            g1 / g2
+          } else {
+            g1 - g2
+          }
+        })
+
+        pVal <- (1 + sum(abs(testStat_permuted) >= abs(testStat_observed)))/(nPermutation + 1)
+        # plus ones to make sure p-val is never 0?
+        return(pVal)
+      }
+
+      pval <- permutation_test(values_x, values_y)
+    } else {
+      pval <- NA
+    }
+    return(list(ratio = ratio, pval = pval))
+  }
+
+  # we only need to do the calculation once, because for the ratio that
+  # should just be the inverse and for the permuation the result should
+  # be the same
+
+  combinations <-   data.table::as.data.table(t(combn(identifiers, m = 2)))
+  colnames(combinations) <- c("model", "compare_against")
+
+  combinations[, c("ratio", "pval") := single_comparison(model, compare_against,
+                                                         permutation_test),
+               by = seq_len(NROW(combinations))]
```

Comment:
is .N not support here? I assume not. New to m.

Reply:
no idea... copied that bit from SO :-D 
not sure what the appropriate way to do rowwise is in data.table...

Reply:
Yes it all looks like it makes sense to me (though I think I would have done it with an I'd car or detecting column names that were not the one I wanted to act on). I was just wondering if the .N row counter built into data.table could be used here.